src/utils/logger.h

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright 2021 John Mifsud <imabuddha@gmail.com>
+ *
+ * This file is part of Touchégg.
+ *
+ * Touchégg is free software: you can redistribute it and/or modify it under the
+ * terms of the GNU General Public License  as  published by  the  Free Software
+ * Foundation,  either version 3 of the License,  or (at your option)  any later
+ * version.
+ *
+ * Touchégg is distributed in the hope that it will be useful,  but  WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+ * A PARTICULAR PURPOSE.  See the  GNU General Public License  for more details.
+ *
+ * You should have received a copy of the  GNU General Public License along with
+ * Touchégg. If not, see <http://www.gnu.org/licenses/>.
+ */
+#ifndef UTILS_LOGGER_H_
+#define UTILS_LOGGER_H_
+
+#include <string>
+
+/**
+ * Singleton class to provide simple logging functionality.
+ */
+class Logger {
```

Comment:
It feels like `Logger::obj().level <<` is a little bit verbose. Maybe, we could copy how `std::cout` works and just add a few methods to a namespace, something like...

```cpp
namespace log {
  void configureLogger(...options...);

  LogLevel error = LogLevel::ERROR;
  LogLevel warning = LogLevel::WARNING;
  LogLevel info = LogLevel::INFO;
  LogLevel debug = LogLevel::DEBUG;
}
```

So we could do something like:

```cpp
log::info << "Info message" << std::endl;
log::error << "A terrible error"  << std::endl;
```

What do you think?

Reply:
Yes the `Logger::obj().level <<` is a bit verbose, which is why I create a var for log in funcs that do multiple msgs. In those use cases it's 1 char smaller than your suggestion: `log.info`.

TBH I'm not 100% happy with the current implementation either; the class member level vars are a clunky hack. Your suggestion is a good one, and eliminates them.

**>> What purpose would the configureLogger func serve?**

### _Personal opinion rant:_
As you know I had a much simpler mod in mind, just adding another touchegg cmd line flag or two & using them to suppress gesture message output with if statements. Your concept of a singleton used with << overloading made it much more time consuming to implement.

**Using a singleton** to store the global logging options & suppress msgs as needed **is cleaner and the right way to do it**. However, << overloading for this use case is a **PITA**, mainly due to c++. The << func must be a non-member friend of the singleton otherwise the first param would be the instance, giving no way to know which log level the msg is for.

`std::end` is another crazy c++ thing. It's not clear to me what benefits it has over a simple `\n`. Not only that but the way it's implemented requires adding a second << function. 😝

Be that as it may, it's your prj so I fought with c++ and got it working. Not being a c++ expert for the last ~20 years (used to have [and I actually read!] all the books: ARM, the one before the ARM, the giant white bible size book, etc.) Much of the effort was due to my lack of recent experience with it.

Reply:
Unfortunately the namespace will have to be different from 'log' of we get this error:
```
‘namespace log { }’ redeclared as different kind of symbol
previous declaration ‘double log(double)’
```

Reply:
So the name will be `tlg` for touchegg log. Usage like `tlg::info << "hello"`